src/sys/windows/selector.rs

```diff
@@ -371,14 +411,233 @@ impl Selector {
+
+    use crate::poll;
+    use crate::{Registry, Token};
+
+    const OVERLAPPED_CANARY: u32 = 0x5e219f3a;
+
+    /// Holds a set of actions that the sending IOCP client is ready to do.
+    ///
+    /// These can be combined into one `Readiness` instance by using the bitwise or operator '|'.
+    /// However, there is no "not ready" state, i.e. and empty `Readiness`. To indicate the absence of
+    /// any ready operation, use Option<Readiness> and return a None value.
+    #[derive(Debug, Clone, Copy, Eq, PartialEq)]
+    pub struct Readiness {
+        flags: NonZeroU32,
+    }
+
+    impl Readiness {
+        /// Readiness that indicates readiness to read
+        pub const READ: Readiness = Readiness { flags: unsafe { NonZeroU32::new_unchecked(0b_0000_0001) } };
+        /// Readiness that indicates readiness to write
+        pub const WRITE: Readiness = Readiness { flags: unsafe { NonZeroU32::new_unchecked(0b_0000_0010) } };
+
+        /// Checks whether the provided readiness is fulfilled
+        pub fn is_set(&self, readiness: Readiness) -> bool {
+            (self.flags.get() & readiness.flags.get()) == readiness.flags.get()
+        }
+
+        pub(crate) fn into_event(self, token: Token) -> Event {
+            let readable = if self.is_set(Self::READ) { afd::POLL_RECEIVE } else { 0 };
+            let writable = if self.is_set(Self::WRITE) { afd::POLL_SEND } else { 0 };
+            let flags = readable | writable;
+            debug_assert!(flags != 0);
+            Event {
+                data: token.0 as u64,
+                flags
+            }
+        }
+    }
+
+    impl BitOr for Readiness {
+        type Output = Readiness;
+
+        fn bitor(self, rhs: Self) -> Self::Output {
+            // Since neither self nor rhs can be zero (assuming absence of transmuted Readinesses), it's
+            // safe to assume that the result is also nonzero.
+            let flags = unsafe {
+                NonZeroU32::new_unchecked(self.flags.get() | rhs.flags.get())
+            };
+            Readiness { flags }
+        }
+    }
+
+    /// Callback trait for completed IOCP operations.
+    ///
+    /// This trait may be implemented to handle completed IOCP operations. The implementing object may
+    /// signal the readiness for certain operations to the upper layer by returning a `Readiness` value
+    /// that is then converted into `mio::Event` and delivered to the user.
+    ///
+    /// This trait is implemented for closures as well.
+    pub trait CompletionCallback {
+        /// Called on completion of the associated operation.
+        ///
+        /// The callback will be called once the operation that it is bound to via a call to
+        /// `Overlapped::new` has been completed and the IOCP is signalled. By returning a `Readiness`
+        /// value, the callback may deliver a `mio::Event` to the user.
+        fn complete_operation(&mut self, entry: &OVERLAPPED_ENTRY) -> Option<Readiness>;
```

Comment:
Personally not the biggest fan out exposing types from different crates, especially when those crates don't have a stable API.

Reply:
I don't think that this happens here: This method resembles the old fn(&OVERLAPPED_ENTRY) from mio 0.6, so except from the OVERLAPPED_ENTRY (which is a stable Windows API struct) we do not expose other types. But maybe I misunderstand your comment.